source/common/ssl/context_impl.cc

```diff
@@ -510,9 +520,31 @@ bssl::UniquePtr<SSL> ClientContextImpl::newSsl() const {
     SSL_set_renegotiate_mode(ssl_con.get(), ssl_renegotiate_freely);
   }
 
+  if (max_session_keys_ > 0) {
+    absl::WriterMutexLock l(&session_keys_mu_);
+    if (!session_keys_.empty()) {
+      SSL_SESSION* session = session_keys_.front().get();
+      SSL_set_session(ssl_con.get(), session);
+      // Remove single-use (TLS v1.3) session key after first use.
+      if (SSL_SESSION_should_be_single_use(session)) {
+        session_keys_.pop_front();
+      }
+    }
+  }
+
   return ssl_con;
 }
 
+int ClientContextImpl::newSessionKey(SSL_SESSION* session) {
```

Comment:
I think we are safe for multi cert work, given that the client contexts will continue to have a single cert, but could you comment here on whether in the future, if we support multiple client certs, whether anything needs to change?

Reply:
I don't think that we'll ever support multiple client certificates that can affect sessions, since client certificates are not revalidated during session resumption by the server.

In any case, this would be covered by https://github.com/envoyproxy/envoy/issues/5073.